File: pyg4ometry/pygeometry/cpgeant4/solid/EllipticalTube.cpp
#include "EllipticalTube.h"
#include <cmath>

void CSGMesh::EllipticalTubeAppendVertex(std::vector<Vertex*> vertices,double theta,double z,double dx,double dy,Vector* norm){
  double x = dx*cos(theta);
  double y = dy*sin(theta);
  Vector* d = new Vector(x,y,z);
  Vector* n;
  if(!norm){
    n = d;
  }
  else{
    n = norm;
  }
  return vertices.push_back(new Vertex(d,n));
}

CSG* CSGMesh::ConstructEllipticalTube(double pDx,double pDy,double pDz,int slices,int stacks){
  double sz = -pDz;
  double dz = (2.0*pDz)/stacks;
  double dTheta = (2.0*M_PI)/slices;

  std::vector<Polygon*> polygons;
  for(int j0=0;j0<slices;j0++){
    double j1 = j0 + 0.5;
    double j2 = j0 + 1.0;
    for(int i0=0;i0<stacks;i0++){
      double i1 = i0 + 0.5;
      double i2 = i0 + 1.0;
      std::vector<Vertex*> verticesN;
      EllipticalTubeAppendVertex(verticesN, i1 * dTheta, j1 * dz + sz,pDx,pDy,NULL);
      EllipticalTubeAppendVertex(verticesN, i2 * dTheta, j2 * dz + sz,pDx,pDy,NULL);
      EllipticalTubeAppendVertex(verticesN, i0 * dTheta, j2 * dz + sz,pDx,pDy,NULL);
      polygons.push_back(new Polygon(verticesN,NULL));
      std::vector<Vertex*> verticesS;
      EllipticalTubeAppendVertex(verticesS, i1 * dTheta, j1 * dz + sz,pDx,pDy,NULL);
      EllipticalTubeAppendVertex(verticesS, i0 * dTheta, j0 * dz + sz,pDx,pDy,NULL);
      EllipticalTubeAppendVertex(verticesS, i2 * dTheta, j0 * dz + sz,pDx,pDy,NULL);
      polygons.push_back(new Polygon(verticesS,NULL));
      std::vector<Vertex*> verticesW;
      EllipticalTubeAppendVertex(verticesW, i1 * dTheta, j1 * dz + sz,pDx,pDy,NULL);
      EllipticalTubeAppendVertex(verticesW, i0 * dTheta, j2 * dz + sz,pDx,pDy,NULL);
      EllipticalTubeAppendVertex(verticesW, i0 * dTheta, j0 * dz + sz,pDx,pDy,NULL);
      polygons.push_back(new Polygon(verticesW,NULL));
      std::vector<Vertex*> verticesE;
      EllipticalTubeAppendVertex(verticesE, i1 * dTheta, j1 * dz + sz,pDx,pDy,NULL);
      EllipticalTubeAppendVertex(verticesE, i2 * dTheta, j0 * dz + sz,pDx,pDy,NULL);
      EllipticalTubeAppendVertex(verticesE, i2 * dTheta, j2 * dz + sz,pDx,pDy,NULL);
      polygons.push_back(new Polygon(verticesE,NULL));

    }
  }

  for(int i0=0;i0<slices;i0++){
    double i1 = i0 + 1;
    std::vector<Vertex*> vertices1;
    EllipticalTubeAppendVertex(vertices1,i0*dTheta,sz,pDz,pDy,new Vector(0.0,0.0,1.0));
    EllipticalTubeAppendVertex(vertices1,0,sz,0,0,new Vector(0.0,0.0,1.0));
    EllipticalTubeAppendVertex(vertices1,i1*dTheta,sz,pDz,pDy,new Vector(0.0,0.0,1.0));
    polygons.push_back(new Polygon(vertices1,NULL));

    std::vector<Vertex*> vertices2;
    EllipticalTubeAppendVertex(vertices2,i1*dTheta,stacks*dz+sz,pDx,pDy,new Vector(0.0,0.0,-1.0));
    EllipticalTubeAppendVertex(vertices2,0,stacks*dz+sz,0,0,new Vector(0.0,0.0,-1.0));
    EllipticalTubeAppendVertex(vertices2,i0*dTheta,stacks*dz+sz,pDx,pDy,new Vector(0.0,0.0,-1.0));
    polygons.push_back(new Polygon(vertices2,NULL));
  }

  return CSG::fromPolygons(polygons);
}

File: src/pyg4ometry/geant4/solid/EllipticalTube.py
from ... import config as _config

from .SolidBase import SolidBase as _SolidBase

if _config.meshing == _config.meshingType.pycsg:
    from ...pycsg.core import CSG as _CSG
    from ...pycsg.geom import Vector as _Vector
    from ...pycsg.geom import Vertex as _Vertex
    from ...pycsg.geom import Polygon as _Polygon
elif _config.meshing == _config.meshingType.cgal_sm:
    from ...pycgal.core import CSG as _CSG
    from ...pycgal.geom import Vector as _Vector
    from ...pycgal.geom import Vertex as _Vertex
    from ...pycgal.geom import Polygon as _Polygon

import logging as _log

import numpy as _np


class EllipticalTube(_SolidBase):
    """
    Constructs a tube of elliptical cross-section.

    :param name: name of the solid
    :type name:  str
    :param pDx:  length in x
    :type pDx:   float, Constant, Quantity, Variable, Expression
    :param pDy:  length in y
    :type pDy:   float, Constant, Quantity, Variable, Expression
    :param pDz:  length in z
    :type pDz:   float, Constant, Quantity, Variable, Expression
    :param registry: for storing solid
    :type registry: Registry
    :param lunit: length unit (nm,um,mm,m,km) for solid
    :type lunit: str
    :param nslice: number of phi elements for meshing
    :type nslice: int
    :param nstack: number of theta elements for meshing
    :type nstack: int

    """

    def __init__(
        self,
        name,
        pDx,
        pDy,
        pDz,
        registry,
        lunit="mm",
        nstack=None,
        nslice=None,
        addRegistry=True,
    ):
        super().__init__(name, "EllipticalTube", registry)

        self.pDx = pDx
        self.pDy = pDy
        self.pDz = pDz
        self.lunit = lunit
        self.nslice = nslice if nslice else _config.SolidDefaults.EllipticalTube.nslice
        self.nstack = nstack if nstack else _config.SolidDefaults.EllipticalTube.nstack

        self.dependents = []

        self.varNames = ["pDx", "pDy", "pDz"]
        self.varUnits = ["lunit", "lunit", "lunit"]

        if addRegistry:
            registry.addSolid(self)

    def __repr__(self):
        return f"EllipticalTube : {self.name} {self.pDx} {self.pDy} {self.pDz}"

    def __str__(self):
        return f"EllipticalTube : name={self.name} dx={float(self.pDx)} dy={float(self.pDy)} dz={float(self.pDz)}"

    def mesh(self):
        """new meshing based of Tubs meshing"""

        _log.info("ellipticaltube.antlr>")

        from ...gdml import Units as _Units

        luval = _Units.unit(self.lunit)

        pDx = self.evaluateParameter(self.pDx) * luval
        pDy = self.evaluateParameter(self.pDy) * luval
        pDz = self.evaluateParameter(self.pDz) * luval

        _log.info("ellipticaltube.pycsgmesh>")

        polygons = []

        sz = -pDz
        dTheta = 2 * _np.pi / self.nslice
        slices = self.nslice

        pRMin = 0

        for i in range(0, self.nslice, 1):
            i1 = i
            i2 = i + 1

            p1 = dTheta * i1  # + pSPhi
            p2 = dTheta * i2  # + pSPhi

            ###########################
            # tube ends
            ##########################s
            vEnd = []
            vEnd.append(_Vertex([0, 0, pDz]))
            vEnd.append(_Vertex([pDx * _np.cos(p1), pDy * _np.sin(p1), pDz]))
            vEnd.append(_Vertex([pDx * _np.cos(p2), pDy * _np.sin(p2), pDz]))
            polygons.append(_Polygon(vEnd))

            vEnd = []
            vEnd.append(_Vertex([0, 0, -pDz]))
            vEnd.append(_Vertex([pDx * _np.cos(p2), pDy * _np.sin(p2), -pDz]))
            vEnd.append(_Vertex([pDx * _np.cos(p1), pDy * _np.sin(p1), -pDz]))
            polygons.append(_Polygon(vEnd))

            ###########################
            # Curved cylinder faces
            ###########################
            vCurv = []
            vCurv.append(_Vertex([pDx * _np.cos(p1), pDy * _np.sin(p1), -pDz]))
            vCurv.append(_Vertex([pDx * _np.cos(p2), pDy * _np.sin(p2), -pDz]))
            vCurv.append(_Vertex([pDx * _np.cos(p2), pDy * _np.sin(p2), pDz]))
            vCurv.append(_Vertex([pDx * _np.cos(p1), pDy * _np.sin(p1), pDz]))
            polygons.append(_Polygon(vCurv))

        mesh = _CSG.fromPolygons(polygons)

        return mesh
